Library/Services/Implementation/NfieldSurveySampleService.cs

```diff
@@ -103,10 +107,45 @@ public Task<int> ResetAsync(string surveyId, string respondentKey)
 
             return Client.PutAsJsonAsync<IEnumerable<SampleFilter>>(uri, filters)
                 .ContinueWith(responseMessageTask => responseMessageTask.Result.Content.ReadAsStringAsync().Result)
-                .ContinueWith(stringResult => JsonConvert.DeserializeObject<SampleResetStatus>(stringResult.Result).ResetCount)
+                .ContinueWith(stringResult => JsonConvert.DeserializeObject<BackgroundActivityStatus>(stringResult.Result).ActivityId)
+                .ContinueWith(activityResult => GetActivityResultAsync(activityResult.Result))
+                .Unwrap()
                 .FlattenExceptions();
         }
 
+        /// <summary>
+        /// Recursive method that polls the activity status until it completes.
+        /// </summary>
+        /// <param name="activityId">The id of the activity to wait for.</param>
+        /// <returns></returns>
+        private Task<int> GetActivityResultAsync(string activityId)
+        {
+            return Client.GetAsync(BackgroundActivityUrl(activityId))
+                .ContinueWith(response => response.Result.Content.ReadAsStringAsync())
+                .Unwrap()
+                .ContinueWith(content =>
+                {
+                    var obj = JObject.Parse(content.Result);
+                    var status = obj["Status"].Value<int>();
+
+                    switch (status)
+                    {
+                        case 0: // pending
+                        case 1: // started
+                            Thread.Sleep(millisecondsTimeout: 200);
+                            return GetActivityResultAsync(activityId);
+                        case 2: // succeeded
+                            var tcs = new TaskCompletionSource<int>();
+                            tcs.SetResult(obj["ResetTotal"].Value<int>());
+                            return tcs.Task;
+                        case 3: // failed
+                        default: 
+                            throw new Exception("Reset did not complete successfully");
```

Comment:
Don't throw generic `Exception`.

https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/exceptions/creating-and-throwing-exceptions
> Things to Avoid When Throwing Exceptions

> Do not throw Exception, SystemException, NullReferenceException, or IndexOutOfRangeException intentionally from your own source code.